### PR TITLE
feat(perpetual) add funding rate applier

### LINKED ISSUE #2025: Build Mock Funding Rate Store
**Summary**

Help test Perpetual contract's ability to apply funding rate on token redemption value.

New production contracts to write:
- `FundingRateStoreInterface`

New test contracts to write:
- `MockFundingRateStore`

**Implementation**

Interface should implement the `FundingRateStoreInterface`:
- `getFundingRate() => FixedPoint`

Proposed helper external methods:
- `setFundingRate(FixedPoint newRate) => void`

**Difficulty Score [1-10]**

This is a measure of the amount of developer time this task will require to be completed. Note: this should include design and planning time in addition to implementation time. The scale is not linear:

- 7: ~3 days

### LINKED ISSUE #2030: Build Funding Rate Applier
**Summary**

This contract reads funding rates from a `FundingRateStore` and uses it to continuously update a funding rate multiplier, `cumulativeFundingRateMultiplier`, state variable, similar to how `FeePayer.sol` keeps track of `cumulativeFeeMultiplier` to assess regular fees. The `cumulativeFundingRateMultiplier` comes into play on emergency shutdowns, liquidations, and disputes by modifying token redemption value (TRV).

A `PositionManager` or `Liquidatable` that wants to assess funding rates on TRV should import this contract.

New production contracts to write:
- `FundingRateApplier`

**Implementation**

Proposed construction parameters must also be set up:
```
        ...TBD,
        uint _initialFundingRate, // Starting funding multiplier, e.g. 1.1, 1.0, 0.9, etc.
        address _fpFinderAddress, // fetch financial-product related contracts
        address _timerAddress, // Testable timing contract

```

***Financial Product Finder***
This should point to a `FundingRateStore` that returns the current funding rate, which this contract can use to update its `cumulativeFundingRateMultiplier` value.

**Interface**
- A modifier `updateFunding` should be exposed so that client contracts such as `PositionManager` can update the funding rate multiplier on each contract call based on the funding rate store's latest funding rate.
- `pfc()` should be updated to reflect funding rate effects on profit from corruption.
- Internal methods should be exposed that return the TRV of a synthetic token unit given an oracle price and the current `cumulativeFundingRateMultiplier` value.

**How**

This should be modeled after `FeePayer` in that the `PositionManager` should inherit it, but it will probably be substantially different.

**Difficulty Score [1-10]**

- 8: ~1 week

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

#2030 

**Summary**

Funding rate store represents funding rate as a % of token debt charged per second, added to or subtracted from 100% depending on if the store wants to return a "negative" or "positive" funding rate. For example, a positive funding rate of 5% (longs pay shorts 5% per second) is represented as 1.05. A negative rate of 5% (shorts pay longs 5% per second) is represented as 0.95. This convention is taken so that the interface for users remains consistent in that they use FixedPoint to represent percentages: 1.05 = 1.05e18, and 0.95 = 0.95e18.

The funding rate applier applies the funding rate linearly per second before multiplying it with the current cumulativeFeeMultiplier. For example, a negative funding rate of 5% applied for 5 seconds results in the cumulativeFeeMultiplier being multiplied by (0.05 * 5) = 0.25.

Simple test included to demonstrate feature.

**Issue(s)**

close #2025 
close #2030

### CODE REVIEW

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:90]: drive by: perhaps you're thinking of this already, but shouldn't the store's latest funding rate be applied to the current `cumulativeFundingRateMultuiplier`?

i.e.
- cumulativeFundingRateMultiplier = 1.10
- fundingRateStore.getLatestFundingRateForIdentifier(identifer) = 1.05
- new cumulativeFundingRateMultiplier = 1.10 * 1.05 = 1.155

Comment (chrismaree) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:90]: Ye, this is something I had considered. I believe that we'd only want to update the `cumulativeFundingRateMultiplier` if the funding rate has been updated from the previous value. I.e do the calculation in your comment iff the funding rate has changed.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:135]: Hm this is mostly to be consistent with how timing is done in the rest of the contract. What's nice is that Solidity provides a `now` keyword that returns the current time. I'd recommend we stick with this because it would require a larger refactor

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:133]: This is a good idea, I do think we should move all of the computation into a `pure` function. It would be easier to test in isolation.

**Comment by nicholaspai** [comment]
@UMAprotocol/eng Following @daywiss 's idea, I've refactored the computation of the `effectiveFundingRate` into an internal pure function. This should enable easier testing. Additionally, I've added a `MockFundingRateApplier` contract that we can use to test the two internal methods:

- `_calculateEffectiveFundingRatePerToken`: returns new funding rate multiplier and effective funding rate given a payment period, a latest funding rate, and a starting multiplier
- `_applyEffectiveFundingRatePerToken`: takes the results `_calculateEffectiveFundingRatePerToken` to set new contract state and emit events

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:121]: Yeah, this is something the crypto community goes back and forth on, and I prefer time for anything that relates to interest. People think about interest and time in terms of seconds rather than blocks. If we use blocks to compute interest or funding payments, then the APY in terms of real time changes based on the average block time, which is not ideal IMO.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:163]: nit: could we just remove this case entirely and have this still work as expected?

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:182]: I see what y'all were saying -- the lack of a SignedFixedPoint makes this more complicated than it needs to be. This doesn't necessarily need to change, but I think everything would be more intuitive if the rates came in from the store as positive and negative (centered around 0) and we were able to do this in a single case that handled both positive and negative (which would be pretty trivial if we could use negative numbers).

Probably something we should consider (but outside the scope of this PR IMO).

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:150]: nit: I think `PerToken` in these names is consistent with how we named them in the store, but do you think this is more effective than `fundingRatePerSecond`? I just sometimes find these long names confusing (I feel the same way when I looks back at the Store code). Just wanted to get y'all's thoughts. Don't have any other alternatives in mind, but definitely open to other, more intuitive naming schemes if y'all have any in mind.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:156]: +1

File: packages/core/contracts/financial-templates/test/MockFundingRateApplier.sol
Comment (mrice32) [inline packages/core/contracts/financial-templates/test/MockFundingRateApplier.sol:9]: tiny tiny nit: is this a mock? I'm actually not sure what the exact definition of a mock is (so I totally could be wrong), but elsewhere we've just called these contracts `FundingRateApplierTest` rather than a mock since they inherit from and give access to the internals/private functions of a production contract rather than "faking" or "mocking" its behavior for use by other contracts.

See https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/implementation/test/ResultComputationTest.sol and https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/implementation/test/GovernorTest.sol.

File: packages/core/test/financial-templates/common/FundingRateApplier.js
Comment (mrice32) [inline packages/core/test/financial-templates/common/FundingRateApplier.js:61]:
```suggestion
    // Effective Rate: 0.0015 * 20 = 0.03, funding rate is positive so add 1 => 1.03
```

File: packages/core/test/financial-templates/common/FundingRateApplier.js
Comment (mrice32) [inline packages/core/test/financial-templates/common/FundingRateApplier.js:50]:
```suggestion
    // Effective Rate: 0.0015 * 20 = 0.03, funding rate is positive so add +1 => 1.03
```

File: packages/core/test/financial-templates/common/FundingRateApplier.js
Comment (mrice32) [inline packages/core/test/financial-templates/common/FundingRateApplier.js:72]:
```suggestion
    // Effective Rate: 0.0015 * 20 = 0.03, funding rate is negative so subtract from 1 => 0.97
```

File: packages/core/test/financial-templates/common/FundingRateApplier.js
Comment (mrice32) [inline packages/core/test/financial-templates/common/FundingRateApplier.js:83]:
```suggestion
    // Effective Rate: 0 * 20 = 0, funding rate is neutral so => 1
```

**Review by mrice32 — COMMENTED** [review]
Looks awesome! A few fairly minor comments and a minor suggestion in the test.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:34]: I did this because we previously kept non-oracle related contracts outside of the Finder, so I wasn't sure if we wanted to re-use the "oracle" finder for these non-oracle contracts

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:182]: Downside to this is that the users of the Store would be dealing with `int256`, including voters, unless we converted from `int256 -> FixedPoint` within the Store. WDYT?

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:150]: I'm good with that, `perToken` is probably implied anyways

File: packages/core/contracts/financial-templates/test/MockFundingRateApplier.sol
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/test/MockFundingRateApplier.sol:9]: Good call, it's not a mock its just a test wrapper contract

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: Yep, fixed!

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: I think voters already deal with `int256` (see [here](https://github.com/UMAprotocol/protocol/blob/b3f243a37204bd71fa466f4da5e15754514b8063/packages/core/contracts/oracle/interfaces/OracleInterface.sol#L33)). Regardless, we could use `int256` or we could just build a `SignedFixedPoint` that's used everywhere (including in the interfaces). In general, though, the voters will have to return an `int256` that gets converted to a `FixedPoint` or a `SignedFixedPoint` in the contract.

Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: Thoughts?

Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: Hm this is a good point. I think we should have a couple discussions:
- pros/cons to building SignedFixedPoint
- whether voters should vote in an int256 or a SignedFixedPoint or a FixedPoint

Regardless I think we should push this decision to a subsequent PR. Wdyt? This PR mainly sets up the architecture of the FundingRateApplier but we should decide before we merge in #2084

Comment (nicholaspai) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: Thinking about this some more. I think there is an open question how the FundingRateStore should represent funding rates: as `int256`, `FixedPoint` or `SignedFixedPoint`. But, I think we should keep storing the `cumulativeFundingRateMultiplier` as a `FixedPoint` centered around 1. So, I think this PR can be unchanged.

**Review by nicholaspai — APPROVED** [review]

**Review by daywiss — APPROVED** [review]

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: Yes 100%. We should have this discussion, but do it separately in its own issue/space. No changes needed right now.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol
Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:34]: Seems reasonable for now. We can think more about it down once everything is piped together.

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**
